Move to Rails API instead of Grape

wwahammy opened the pull request:
**NOTE: DO NOT discuss internal CommitChange information in your PR; this PR will be public.
Link back to the issue in the Tix repo when you need to do that.**

Since Grape is a pain to get working on Ruby 3, I migrated that API to a standard Rails controller.

For the most part, I kept the the API similar so it works properly with our Javascript. The key differences:

* I stripped out the v1 after /api in the path. So now it's `/api/nonprofits`. 
* I changed the error json returned the create path from the weird style we had before to a standard ActiveModel::Error style. One of the advantages of TypeScript is that it made it easier to find out where the previous weird error style was.

As part of this, I reorganized the complex register nonprofit code by using [yaaf](https://github.com/rootstrap/yaaf). It's not without its faults but it does make it pretty easy to use the form object pattern in a pretty natural ActiveModel-y way. It also is like 34 lines of code so it's not terribly hard to override things if needed.

Since a YAAF::Form is like any ActiveModel object, I used multiple levels of Forms. I think that's a reasonable way to put each of these pieces into their own discrete, testable parts.

Note: the only place we still used grape was for nonprofit registration. So that's the only place that needed to be tested. It seems to work but it'll definitely require more testing once it's on staging.

wwahammy committed:
Add yaaf

wwahammy committed:
Setup forms

wwahammy committed:
Minor tweak

wwahammy committed:
Remove no-longer used grape and its friends!

wwahammy committed:
Move spec to new location

wwahammy committed:
Remove unused method in spec

wwahammy committed:
Remove unused code in the api/nonprofits_spec

wwahammy committed:
Merge branch 'remove-unused-helper' into move-to-rails-api

wwahammy committed:
Merge branch 'update-stripe-gem-to-5' into move-to-rails-api

# Conflicts:
#	Gemfile.lock

wwahammy committed:
Merge branch 'update-stripe-gem-to-5' into move-to-rails-api

wwahammy committed:
Remove no longer needed grape translations

wwahammy committed:
Merge branch 'update-stripe-gem-to-5' into move-to-rails-api

wwahammy committed:
Merge branch 'update-stripe-gem-to-5' into move-to-rails-api

wwahammy committed:
Merge branch 'update-stripe-gem-to-5' into move-to-rails-api

wwahammy committed:
Merge branch 'update-stripe-gem-to-5' into move-to-rails-api

wwahammy committed:
Cleanup NonprofitForm spec

wwahammy committed:
Cleanup in RegisterNonprofitForm spec

wwahammy commented on app/forms/application_form.rb:
```diff
@@ -1,5 +1,4 @@
 # License: AGPL-3.0-or-later WITH Web-Template-Output-Additional-Permission-3.0-or-later
```
I don't know why Git thinks this was a renamed file but whatevs

wwahammy commented on app/forms/register_nonprofit_form/stripe_account_form.rb:
```diff
@@ -0,0 +1,50 @@
+    false
+  end
+
+  def save!(options = {})
+    super(options)
+
+    if nonprofit.persisted? &&  nonprofit.stripe_account_id.blank?
+      stripe_account_id = StripeAccountUtils.create(nonprofit)
+      unless stripe_account_id.present?
+        add_couldnt_create_stripe_error
+        raise ActiveModel::ValidationError, self
+      end
+      nonprofit.reload # StripeAccountUtils.create is weird and we need to reload after
+    else
+      add_couldnt_create_stripe_error
+      raise ActiveModel::ValidationError, self
+    end
+  end
```
Based on Yaaf

wwahammy commented on app/views/api/nonprofits/create.json.jbuilder:
```diff
@@ -1,4 +1,2 @@
 # License: AGPL-3.0-or-later WITH Web-Template-Output-Additional-Permission-3.0-or-later
```
Not a rename Git but once again, whatevs

wwahammy committed:
Do some basic cleanup of specs

wwahammy committed:
Remove reloader for the grape api

wwahammy committed:
Merge branch 'update-stripe-gem-to-5' into move-to-rails-api

wwahammy committed:
Merge branch 'update-stripe-gem-to-5' into move-to-rails-api

wwahammy committed:
Merge branch 'update-stripe-gem-to-5' into move-to-rails-api

wwahammy committed:
Merge branch 'update-stripe-gem-to-5' into move-to-rails-api

wwahammy committed:
Merge branch 'update-stripe-gem-to-5' into move-to-rails-api

wwahammy committed:
Merge branch 'update-stripe-gem-to-5' into move-to-rails-api

wwahammy committed:
Merge branch 'update-stripe-gem-to-5' into move-to-rails-api

wwahammy commented on javascripts/api/api/NonprofitApi.ts:
```diff
@@ -44,7 +44,7 @@ export class NonprofitApi {
      */
     public getNonprofitId(id: number, extraJQueryAjaxSettings?: JQueryAjaxSettings): JQueryPromise<models.Nonprofit > {
-        let localVarPath = this.basePath + '/nonprofit/{id}'.replace('{' + 'id' + '}', encodeURIComponent(String(id)));
+        let localVarPath = this.basePath + '/nonprofits/{id}'.replace('{' + 'id' + '}', encodeURIComponent(String(id)));
```
Use the standard Rails route names

wwahammy committed:
Merge branch 'update-stripe-gem-to-5' into move-to-rails-api

wwahammy committed:
Merge branch 'update-stripe-gem-to-5' into move-to-rails-api

wwahammy committed:
Merge branch 'update-stripe-gem-to-5' into move-to-rails-api

wwahammy committed:
Merge branch 'update-stripe-gem-to-5' into move-to-rails-api

wwahammy committed:
Merge branch 'update-stripe-gem-to-5' into move-to-rails-api

wwahammy committed:
Merge branch 'update-stripe-gem-to-5' into move-to-rails-api

wwahammy committed:
Merge branch 'update-stripe-gem-to-5' into move-to-rails-api

caseyhelbling commented on spec/requests/api/nonprofits_spec.rb:
```diff
@@ -17,35 +19,35 @@ def expect_validation_errors(actual, input)
-                h(params: ["nonprofit[state_code]"], messages: gr_e("presence", "blank")),
-                h(params: ["nonprofit[city]"], messages: gr_e("presence", "blank")),
+            {
+                "nonprofit[name]" => ["can't be blank"],
```
Looks so much better!

caseyhelbling commented on app/forms/application_form.rb:
```diff
@@ -1,5 +1,4 @@
 # License: AGPL-3.0-or-later WITH Web-Template-Output-Additional-Permission-3.0-or-later
-class Houdini::Api < Grape::API
-  format :json
-  mount Houdini::V1::Api => '/v1'
```
Confirming we don't have any external clients using this `/v1` API...

wwahammy replied:
Nope, the only place this API is currently used is for Nonprofit registration

caseyhelbling committed:
I ran standrdrb on the new form classes and specs in an attempt to
preemptively standardize them. As it was, the extra whitespace and
linting things were distracting and making it hard to review the
substance of the PR.

caseyhelbling committed:
Fix zeitwerk:check

wwahammy committed:
Merge branch 'update-stripe-gem-to-5' into move-to-rails-api

caseyhelbling committed:
Update per code review

caseyhelbling committed:
Revert rubcop recommendation

caseyhelbling committed:
Expect

caseyhelbling committed:
Merge branch 'move-to-rails-api' into feature/clh/move-to-rails-api-with-linting

wwahammy committed:
Apply suggestions from code review

caseyhelbling committed:
Merge pull request #1098 from CommitChange/feature/clh/move-to-rails-api-with-linting

Friendly linting PR

wwahammy committed:
Merge branch 'update-stripe-gem-to-5' into move-to-rails-api

caseyhelbling commented on app/controllers/api/nonprofits_controller.rb:
```diff
@@ -0,0 +1,39 @@
+    :city,
+    :email,
+    :phone]
+    )[:nonprofit] || {}
```
Formatting nit: 
```
    params.permit( nonprofit: [:name, 
      :website,
      :zip_code,
      :state_code,
      :city,
      :email,
      :phone])[:nonprofit] || {}
```

wwahammy replied:
Nice, please submit a change.

caseyhelbling replied:
(like in another PR?)

caseyhelbling replied:
I gotchu - one sec.

wwahammy replied:
You can push to this one or do a code suggestion.

caseyhelbling commented on app/controllers/api/nonprofits_controller.rb:
```diff
@@ -0,0 +1,39 @@
+    params.permit(user: [ :name,
+    :email,
+    :password,
+    :password_confirmation])[:user] || {}
```
Formatting Nit:
```
    params.permit(user: [:name,
      :email,
      :password,
      :password_confirmation])[:user] || {}
```

wwahammy replied:
Nice, please submit a change.

caseyhelbling commented on app/forms/register_nonprofit_form/nonprofit_form.rb:
```diff
@@ -0,0 +1,65 @@
+      phone: phone,
+      state_code: state_code,
+      website: website,
+      zip_code: zip_code
```
Is it ruby 3 where we can start leaving off the variables in the hash param syntax? 

```
OnboardAccounts.set_nonprofit_defaults(city:, email:, name:, phone:, state_code:, website:, zip_code: ))
```

wwahammy replied:
3.1, I think.

caseyhelbling commented on spec/forms/register_nonprofit_form_spec.rb:
```diff
@@ -0,0 +1,144 @@
+      user_attributes: {name: "Name", email: "em@em.com", password: "12345678", password_confirmation: "12345678"}
+    }
+
+    allow_any_instance_of(SlugNonprofitNamingAlgorithm)
```
Did I miss another allow/expect?

wwahammy replied:
Oops, I think so.

caseyhelbling reviewed (commented):
Looks pretty good. Couple minor questions/comments.

caseyhelbling committed:
Format cleanup and allows to expect

wwahammy committed:
Merge pull request #1103 from CommitChange/feature/move-to-rails-api-2

Format cleanup and allows to expect

wwahammy committed:
Remove unneeded #to_hash call

wwahammy committed:
Merge branch 'supporter_level_goal' into move-to-rails-api

caseyhelbling reviewed (approved):
Sweet!
